### PR TITLE
[SPARK-14387][SPARK-19459][SQL] Enable Hive-1.x ORC compatibility with spark.sql.hive.convertMetastoreOrc

### PR DESCRIPTION
**Pull request by dongjoon-hyun**
## What changes were proposed in this pull request?

This PR includes #14471 to enable `spark.sql.hive.convertMetastoreOrc` for SPARK-19459 (ORC tables cannot be read when they contain char/varchar columns). All credits should go to @rajeshbalamohan .

For SPARK-19459, the padding is handled by Hive-side `HiveCharWritable` via `HiveBaseChar.java` on read. So, ORCFileFormat can handle this unlikely from ParquetFileFormat. For Parquet, please see SPARK-21997.

## How was this patch tested?

Pass the newly added test cases.

### CODE REVIEW

**Comment by gatorsmile** [comment]
I think this is not a fix. What is the root cause?

**Comment by vanzin** [comment]
Won't this cause Hive's parquet code to be used by default instead of Spark's? I'm not sure that's what we want.

**Comment by dongjoon-hyun** [comment]
Thank you for review, @gatorsmile and @vanzin . Yes, this is not a way where Apache Spark is heading. It's slow.

Here, I wanted to raise this issue of `convertMetastoreXXX` things since `convertMetastoreXXX` seems not to be tested completly for both ORC/Parquet. Especially, as you see here, 14 failures means the test cases depends on implicitly on that option. I think we need to test for both `true` and `false` for all those cases.

**Comment by dongjoon-hyun** [comment]
I'm not sure where I hit this before. But, I updated 'WHERE' clause examples. Maybe, there was discussion on this difference.

**Comment by vanzin** [comment]
> I wanted to raise this issue of convertMetastoreXXX things since convertMetastoreXXX seems not to be tested completly

That's fine, but that's not what your change is proposing. If you look at the test failures, at least the one I looked at is because the wrong exception is being thrown, not because the feature is broken. That doesn't excuse the fact that it's not tested, but maybe the situation is not that dire.

But as Xiao has mentioned, you're not fixing the problem you describe in the change summary; you haven't even root caused the problem. You just found out that using Hive classes to read the parquet data works around the problem, but that's not an acceptable fix for the problem - it's just a workaround that those affected by the issue can use.

**Comment by dongjoon-hyun** [comment]
I agree with you. Okay. I'll refocus this.

**Comment by dongjoon-hyun** [comment]
@gatorsmile and @vanzin .

I'm comparing with ORC now. Previously, ORC fails with another reason. I'll make another PR for that. I found that #14471 is enough for ORC.

In case of ORC, ORC itself handles truncations on write. The padding is handled by Hive side `HiveCharWritable` via [HiveBaseChar.java](https://github.com/apache/hive/blob/master/common/src/java/org/apache/hadoop/hive/common/type/HiveBaseChar.java#L57) on read. In case of Parquet, I guess Parquet is the same, but there is no such a padding logic like HiveCharWritable in Spark.

**Comment by dongjoon-hyun** [comment]
Since this PR is invalid, I'll reuse this PR instead of creating new one.

File: sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala [hunk 1]
```diff
@@ -187,8 +188,12 @@ abstract class OrcSuite extends QueryTest with TestHiveSingleton with BeforeAndA
            |STORED AS orc
            |LOCATION '$uri'""".stripMargin)
       val result = Row("a", "b         ", "c", Seq("d  "))
-      checkAnswer(spark.table("hive_orc"), result)
-      checkAnswer(spark.table("spark_orc"), result)
+      Seq("false", "true").foreach { value =>
+        withSQLConf(HiveUtils.CONVERT_METASTORE_ORC.key -> value) {
```
Comment (dongjoon-hyun) [inline sql/hive/src/test/scala/org/apache/spark/sql/hive/orc/OrcSourceSuite.scala]: This is a test case for ORC file format. Based on #14471, I'm enabling this.
For Parquet, I think we can proceed separately if ORC is finished.

**Pull request closed without merging**